exchanges/stream/websocket_test.go

```diff
@@ -0,0 +1,956 @@
+		testData := &testCases[i]
+		t.Run(testData.WC.ExchangeName, func(t *testing.T) {
+			if testData.WC.ProxyURL != "" && !useProxyTests {
+				t.Skip("Proxy testing not enabled, skipping")
+			}
+			err := testData.WC.Dial(&dialer, http.Header{})
+			if err != nil {
+				if testData.Error != nil && strings.Contains(err.Error(), testData.Error.Error()) {
+					return
+				}
+				t.Fatal(err)
+			}
+			err = testData.WC.SendJSONMessage(Ping)
+			if err != nil {
+				t.Error(err)
+			}
+			err = testData.WC.SendRawMessage(websocket.TextMessage, []byte(Ping))
+			if err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+// TestSendMessageWithResponse logic test
+func TestSendMessageWithResponse(t *testing.T) {
+	wc := &WebsocketConnection{
+		Verbose:          true,
+		URL:              "wss://echo.websocket.org",
+		ResponseMaxLimit: time.Second * 5,
+		Match:            NewMatch(),
+	}
+	if wc.ProxyURL != "" && !useProxyTests {
+		t.Skip("Proxy testing not enabled, skipping")
+	}
+
+	err := wc.Dial(&dialer, http.Header{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	go readMessages(wc, t)
+
+	request := testRequest{
+		Event: "subscribe",
+		Pairs: []string{currency.NewPairWithDelimiter("XBT", "USD", "/").String()},
+		Subscription: testRequestData{
+			Name: "ticker",
+		},
+		RequestID: wc.GenerateMessageID(false),
+	}
+
+	_, err = wc.SendMessageReturnResponse(request.RequestID, request)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+// readMessages helper func
+func readMessages(wc *WebsocketConnection, t *testing.T) {
+	timer := time.NewTimer(20 * time.Second)
+	for {
+		select {
+		case <-timer.C:
+			return
+		default:
+			resp := wc.ReadMessage()
+			if resp.Raw == nil {
+				t.Error("connection has closed")
+				return
+			}
+			var incoming testResponse
+			err := json.Unmarshal(resp.Raw, &incoming)
+			if err != nil {
+				t.Error(err)
+				return
+			}
+			if incoming.RequestID > 0 {
+				wc.Match.IncomingWithData(incoming.RequestID, resp.Raw)
+				return
+			}
+		}
+	}
+}
+
+// TestSetupPingHandler logic test
+func TestSetupPingHandler(t *testing.T) {
+	wc := &WebsocketConnection{
+		URL:              "wss://echo.websocket.org",
+		ResponseMaxLimit: time.Second * 5,
+		Match:            NewMatch(),
+		Wg:               &sync.WaitGroup{},
+	}
+
+	if wc.ProxyURL != "" && !useProxyTests {
+		t.Skip("Proxy testing not enabled, skipping")
+	}
+	wc.ShutdownC = make(chan struct{})
+	err := wc.Dial(&dialer, http.Header{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	wc.SetupPingHandler(PingHandler{
+		UseGorillaHandler: true,
+		MessageType:       websocket.PingMessage,
+		Delay:             1000,
+	})
+
+	err = wc.Connection.Close()
+	if err != nil {
+		t.Error(err)
+	}
+
+	err = wc.Dial(&dialer, http.Header{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	wc.SetupPingHandler(PingHandler{
+		MessageType: websocket.TextMessage,
+		Message:     []byte(Ping),
+		Delay:       200,
+	})
+	time.Sleep(time.Millisecond * 500)
+	close(wc.ShutdownC)
```

Comment:
Same with this, the wg.wait was used to check that it the pinghandler exited properly. Removing it doesn't allow that to be verified